lib/ui/channel_buffers.dart

```diff
@@ -6,184 +6,334 @@
 
-  bool _isEmpty(String channel) {
-    final _RingBuffer<_StoredMessage>? queue = _messages[channel];
-    return queue == null || queue.isEmpty;
+  /// Sets the listener for the specified channel.
+  ///
+  /// When there is a listener, messages are sent immediately.
+  ///
+  /// Each channel may have up to one listener set at a time. Setting
+  /// a new listener on a channel with an existing listener clears the
+  /// previous one.
+  ///
+  /// Callbacks are invoked in their own stack frame and
+  /// use the zone that was current when the callback was
+  /// registered.
+  ///
+  /// ## Draining
+  ///
+  /// If any messages were queued before the listener is added,
+  /// they are drained asynchronously after this method returns.
+  ///
+  /// Each message is handled in its own microtask. No messages can
+  /// be queued by plugins while the queue is being drained, but any
+  /// microtasks queued by the handler itself will be processed before
+  /// the next message is handled.
+  ///
+  /// The draining stops if the listener is removed.
+  void setListener(String name, ChannelCallback callback) {
+    final _Channel channel = _channels.putIfAbsent(name, () => _Channel(kDefaultBufferSize));
+    channel.setListener(callback);
   }
 
-  /// Changes the capacity of the queue associated with the given channel.
+  /// Clears the listener for the specified channel.
   ///
-  /// This could result in the dropping of messages if newSize is less
-  /// than the current length of the queue.
-  void _resize(String channel, int newSize) {
-    _RingBuffer<_StoredMessage>? queue = _messages[channel];
-    if (queue == null) {
-      queue = _makeRingBuffer(newSize);
-      _messages[channel] = queue;
-    } else {
-      final int numberOfDroppedMessages = queue.resize(newSize);
-      if (numberOfDroppedMessages > 0) {
-        _Logger._printString('Dropping messages on channel "$channel" as a result of shrinking the buffer size.');
-      }
-    }
+  /// When there is no listener, messages on that channel are queued,
+  /// up to [kDefaultBufferSize] (or the size configured via the
+  /// control channel), and then discarded in a first-in-first-out
+  /// fashion.
+  void clearListener(String name) {
+    final _Channel? channel = _channels[name];
+    if (channel != null)
+      channel.clearListener();
   }
 
   /// Remove and process all stored messages for a given channel.
   ///
   /// This should be called once a channel is prepared to handle messages
   /// (i.e. when a message handler is setup in the framework).
-  Future<void> drain(String channel, DrainChannelCallback callback) async {
-    while (!_isEmpty(channel)) {
-      final _StoredMessage message = _pop(channel)!;
+  ///
+  /// The messages are processed by calling the given `callback`. Each message
+  /// is processed in its own microtask.
+  // TODO(ianh): deprecate once framework uses [setListener].
+  Future<void> drain(String name, DrainChannelCallback callback) async {
```

Comment:
I'm not sure that anyone ever took advantage of the future that was returned here, but now there is no way for a client to know when the draining process is done with the setListener API.  There isn't a place to put it either.  I would rather keep the explicit drain call, maybe replace this with `_drainAsynchronously`?

Reply:
I'd just say launching any asynchronous operation without returning a handle to that operation is a red flag since it hides the fact that something is asynchronous and provides no mechanism for someone to synchronize if they want to.

Reply:
Right now the drain API only has one client, the framework, and it ignores the future.

That said, I think that exposing the end of the drain is a dangerous affordance. The point of this API is to buffer the messages so that when the plugin is ready to receive them, it gets them. The sender has no way to know if the plugin is already listening or not when it sends the message, so by definition the messages it sends cannot depend on whether or not the plugin is listeneing. If we allow messages received before the plugin subscribes to be distinguished to those received after the plugin subscribes, we allow the plugin to behave differently for events sent before and after subscription, but since we just said the sender is not distinguishing, that implies that we are allowing behavior variance along an axis that should have none (i.e., it would be a bug vector).

It also would encourage race conditions. Consider these two cases:

A:
* t=0ms plugin requests buffer size 3
* t=0ms message 1
* t=2ms message 2
* t=3ms plugin subscribes
* t=4ms message 3

B:
* t=0ms plugin requests buffer size 3
* t=0ms message 1
* t=1ms plugin subscribes
* t=2ms message 2
* t=4ms message 3

Why would we want the plugin to react differently to message 2 in the first case vs the second case?

To put it another way, what's the use case for knowing when the drain is complete?

Reply:
I admit it is hypothetical.  Hiding an asynchronous operation in a synchronous signature can cause bugs.  It's not as helpful in Dart though because Dart allows you to drop return values implicitly.

Someday someone might write this code:
```dart
_drain();
clearListener();
```

This is a nonsensical operation because the clearListener would immediately disable the drain before anything happens.

If `_drain` advertises itself as an asynchronous operation by returning a `Future<void>` that mistake is less likely.  I don't know what code some future developer might write but I'd error on the side of giving them best hope of writing correct code.

Reply:
Sorry for not being clear. My medium term plan is to remove the drain API, it just becomes an implicit behavior of setListener. The idea is to remove the concept of draining from the API -- it just behaves exactly as if messages were only sent while the listener was set, and they get processed the same as if there were messages in fact only being sent while the listener was set.

Reply:
Yea, I got that.  I think it's a worse design because it's hiding asynchronous logic inside of the setter.  This is hiding side effects and temporal dependencies which is a liability.  If you are keen on this design naming `setListener` to `setListenerAndDrain` would be an improvement.

Reply:
I don't really understand the concern here. Can you elaborate?

The point of this design is to make the draining entirely transparent, indistinguishable from not draining. There should be no way to distinguish someone connecting to a plugin before the plugin sends messages from connecting after the plugin sends messages. There's no "draining" anymore, conceptually, it's just that the messages weren't delivered until there was a listener.